--- patchqmake.c.orig	2013-03-08 16:24:46.000000000 +0100
+++ patchqmake.c	2013-03-08 16:29:57.000000000 +0100
@@ -13,7 +13,7 @@
 
 #define InFile            argv[1]
 #define TargetPath        argv[2]
-#define ARRAYCOUNT(array) sizeof(array)/sizeof(array[0])
+#define ARRAYCOUNT(array) (sizeof(array)/sizeof(array[0]))
 
 typedef struct
 {
@@ -171,7 +171,8 @@
    char *match;
    char *end;
    char *nextbeg;
-   int i, matches;
+   size_t i;
+   int matches;
    #if defined(__linux__) || defined(__APPLE__)
    FILE *backup = NULL;
    #endif
