--- patchqmake.c.orig	2012-02-12 16:45:42.000000000 +0100
+++ patchqmake.c	2012-02-12 16:49:38.000000000 +0100
@@ -18,6 +18,7 @@
 typedef struct
 {
    char *find;
+   int   findlen;
    char *add;
 } sdata;
 
@@ -27,16 +28,16 @@
    #define PATHSEP   "/"
 #endif
 
-sdata Searches[] = {{"qt_prfxpath=", ""},
-                    {"qt_docspath=", PATHSEP "doc"},
-                    {"qt_hdrspath=", PATHSEP "include"},
-                    {"qt_libspath=", PATHSEP "lib"},
-                    {"qt_binspath=", PATHSEP "bin"},
-                    {"qt_plugpath=", PATHSEP "plugins"},
-                    {"qt_datapath=", ""},
-                    {"qt_trnspath=", PATHSEP "translations"},
-                    {"qt_xmplpath=", PATHSEP "examples"},
-                    {"qt_demopath=", PATHSEP "demos"}
+sdata Searches[] = {{"qt_prfxpath=", 12, ""},
+                    {"qt_docspath=", 12, PATHSEP "doc"},
+                    {"qt_hdrspath=", 12, PATHSEP "include"},
+                    {"qt_libspath=", 12, PATHSEP "lib"},
+                    {"qt_binspath=", 12, PATHSEP "bin"},
+                    {"qt_plugpath=", 12, PATHSEP "plugins"},
+                    {"qt_datapath=", 12, ""},
+                    {"qt_trnspath=", 12, PATHSEP "translations"},
+                    {"qt_xmplpath=", 12, PATHSEP "examples"},
+                    {"qt_demopath=", 12, PATHSEP "demos"}
                    };
 
 
@@ -219,7 +220,7 @@
       for (match = filedata; match < filedata + filesize; match++)
       {
          // See if we've found the string to change
-         if (memcmp(match, Searches[i].find, strlen(Searches[i].find)) == 0)
+         if (memcmp(match, Searches[i].find, Searches[i].findlen) == 0)
          {
             matches++;
 
