--- patchqmake.c.orig	2005-11-08 20:03:13.000000000 +0100
+++ patchqmake.c	2010-10-08 10:51:44.000000000 +0200
@@ -3,7 +3,7 @@
 #include <string.h>
 #if defined(_WIN32)
    #include <windows.h>
-#elif defined(__linux__)
+#elif defined(__linux__) || defined(__APPLE__)
    #include <sys/stat.h>
    #include <sys/mman.h>
 #endif
@@ -18,7 +18,7 @@
 
 #if defined(_WIN32)
    #define PATHSEP   "\\"
-#elif defined(__linux__)
+#elif defined(__linux__) || defined(__APPLE__)
    #define PATHSEP   "/"
 #endif
 
@@ -107,7 +107,7 @@
       CloseHandle(hQMake); 
       return NULL; 
    }
-   #elif defined(__linux__)
+   #elif defined(__linux__) || defined(__APPLE__)
    static FILE *hQMake = NULL;
    static void *map = NULL;
    struct stat filesize;
@@ -168,7 +168,7 @@
    char *end;
    char *nextbeg;
    int i;
-   #if defined(__linux__)
+   #if defined(__linux__) || defined(__APPLE__)
    FILE *backup = NULL;
    #endif
 
@@ -210,7 +210,7 @@
       MapFile(NULL, &filesize);
       return 3;
    }
-   #elif defined(__linux__)
+   #elif defined(__linux__) || defined(__APPLE__)
    backup = fopen(match, "wb");
    if (backup == NULL)
    {
