--- ./qtbase/src/corelib/io/qfsfileengine_unix.cpp.feof	2013-01-29 20:03:02.000000000 +0100
+++ ./qtbase/src/corelib/io/qfsfileengine_unix.cpp	2013-03-06 15:31:09.000000000 +0100
@@ -292,7 +292,7 @@
                     int readByte = 0;
                     do {
                         readByte = fgetc(fh);
-                    } while (readByte == -1 && errno == EINTR);
+                    } while (readByte == EOF && !feof(fh) && errno == EINTR);
                     if (readByte != -1) {
                         *data = uchar(readByte);
                         readBytes += 1;
