--- ./qtbase/src/gui/kernel/qguiapplication.cpp.orig	2015-05-26 14:11:49.000000000 +0200
+++ ./qtbase/src/gui/kernel/qguiapplication.cpp	2015-05-26 16:34:34.000000000 +0200
@@ -1679,7 +1679,7 @@
 {
     QEvent::Type type;
     Qt::MouseButtons stateChange = e->buttons ^ buttons;
-    if (e->globalPos != QGuiApplicationPrivate::lastCursorPosition && (stateChange != Qt::NoButton)) {
+    if (0 && e->globalPos != QGuiApplicationPrivate::lastCursorPosition && (stateChange != Qt::NoButton)) {
         // A mouse event should not change both position and buttons at the same time. Instead we
         // should first send a move event followed by a button changed event. Since this is not the case
         // with the current event, we fake a move-only event that we recurse and process first. This
